File: leetcode/lc-us/cpp/1522.cpp
#include <bits/stdc++.h>

using namespace std;

class Node {
public:
    int val;
    vector<Node *> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node *> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
    int res;

    int dfs(Node *root) {
        if (!root) return 0;
        vector<int> depth;
        int x = 0, y = 0;
        for (auto child: root->children) {
            depth.push_back(dfs(child));
        }
        sort(depth.begin(), depth.end(), greater<>());
        if (!depth.empty()) {
            x = depth.front();
            if (depth.size() > 1) {
                y = depth[1];
            }
        }
        res = max(res, x + y);
        return max(x, y) + 1;
    }

public:
    int diameter(Node *root) {
        res = 0;
        dfs(root);
        return res;
    }
};

File: leetcode/lc-us/py/1522.py
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children if children is not None else []


class Solution:
    def __init__(self):
        self.res = 0

    def diameter(self, root: 'Node') -> int:
        def dfs(node: 'Node') -> int:
            if not node:
                return 0
            a = b = 0
            for child in node.children:
                t = dfs(child)
                if t > a:
                    b, a = a, t
                elif t > b:
                    b = t
            self.res = max(self.res, a + b)
            return max(a, b) + 1

        dfs(root)
        return self.res
